proposals/0000-type-variable-scoping.rst

```diff
@@ -0,0 +1,1417 @@
+of patterns binds the same term variable twice. And we have an accepted ``h``, but this
+``h`` has a shadowed binding for ``x`` -- no equality comparison here. In terms, any variable
+occurrence in a pattern (except in the expression part of a view pattern) is a binding site,
+with no questions asked.
+
+**Visible dependent quantification**. Now, let's consider what happens once we have visible
+dependent quantification (VDQ) in the types of GADT constructors. This is proposed in `#281`_, which
+amends `#402`_ to allow VDQ in GADT constructors (among supporting VDQ more generally). ::
+
+  data VDQ a where
+    MkVDQ :: forall a b -> a -> b -> VDQ a
+
+The type ``VDQ`` is like ``UnivEx``, except that it uses VDQ in its constructor. A pattern might
+look like ``f (MkVDQ a b x y) = ...``. According to the analysis above, we might want variable
+*occurrences* in the first argument of the pattern ``MkVDQ`` (because ``a`` is universal, and
+hence an *input* to the pattern), while we definitely want binding sites for ``x`` and ``y``, the
+ordinary term arguments. Yet, having different binding treatment for ``a`` than for ``x`` and ``y``
+is in violation of the LSPC_, which forbids us from looking at the type of ``MkVDQ`` in making this
+decision.
+
+In order to keep backward compatibility with the current treatment for term-level patterns, we must
+treat ``a`` and ``b`` as *binding sites* not occurrences.
+
+**Visible vs. invisible type arguments**. The VOP_ tells us that the presence or absence of the
+``@`` sign should not affect the binding-site treatment of a region of code. Accordingly, we now
+realize that all type arguments -- whether visible or invisible -- must treat variable mentions
+as binding sites, not occurrences. Here is an example::
+
+  data VOP a b where
+    MkVOP :: forall a -> forall b. a -> b -> VOP a b
+
+Here, we have one visible type argument and one invisible one. A pattern might look like
+``f (MkVOP a @b x y) = ...``. The VOP_ compels us to treat ``a`` and ``b`` identically.
+Furthermore, the LSPC_ compels us to treat ``a`` and ``x``\ /\ ``y`` identically, too.
+Accordingly, all of these must be binding sites -- none can be occurrences.
+
+**Term-level inputs to patterns**. Pattern synonyms currently allow us to declare two
+flavors of input: universal type variables and required constraints. We also can declare
+three forms of output: existential type variables, provided constraints, and normal term-level arguments.
+We're clearly missing something here: term-level input arguments. This proposal does *not*
+address this deficiency, but it seems desirable to leave the door open to such an extension in
+the future. And when we do, the treatment for term-level inputs should be -- in accordance with the
+SUP_ -- the same as the treatment for universal type variables.
+
+**But we want occurrences**. Earlier, we discovered that we sometimes want occurrences
```

Comment:
Did we? It wasn't clear to me reading this that we did. The first example says that it might be "sensible but redundant" to do that, which doesn't seem like a very strong argument. My conclusion at this point in the document was the opposite: I can't see why we need occurences, and instead variables in patterns should always be bindings (agreeing with @nomeata I think).

Reply:
Oh, but I very much _want_ occurrences here! I just admit that they will need dedicated syntax, but not that they are not useful.

Reply:
I don't really care, but since they indeed do need dedicated syntax don't feel motivated to have a strong opinion :).